mm2src/coins/eth/wallet_connect.rs

```diff
@@ -0,0 +1,267 @@
+/// https://docs.reown.com/advanced/multichain/rpc-reference/ethereum-rpc
+use crate::common::Future01CompatExt;
+use crate::Eip1559Ops;
+use crate::{BytesJson, MarketCoinOps, TransactionErr};
+
+use common::log::info;
+use common::u256_to_hex;
+use derive_more::Display;
+use enum_derives::EnumFromStringify;
+use ethcore_transaction::{Action, SignedTransaction};
+use ethereum_types::H256;
+use ethereum_types::{Address, Public, H160, H520, U256};
+use ethkey::{public_to_address, Message, Signature};
+use kdf_walletconnect::WalletConnectOps;
+use kdf_walletconnect::{chain::{WcChainId, WcRequestMethods},
+                        error::WalletConnectError,
+                        WalletConnectCtx};
+use mm2_err_handle::prelude::*;
+use secp256k1::PublicKey;
+use secp256k1::{recovery::{RecoverableSignature, RecoveryId},
+                Secp256k1};
+use std::str::FromStr;
+use web3::signing::hash_message;
+
+use super::{EthCoin, EthPrivKeyPolicy};
+
+// Wait for 60 seconds for the transaction to appear on the RPC node.
+const WAIT_RPC_TIMEOUT_SECS: u64 = 60;
+
+#[derive(Display, Debug, EnumFromStringify)]
+pub enum EthWalletConnectError {
+    UnsupportedChainId(WcChainId),
+    InvalidSignature(String),
+    AccoountMisMatch(String),
+    #[from_stringify("rlp::DecoderError", "hex::FromHexError")]
+    TxDecodingFailed(String),
+    #[from_stringify("ethkey::Error")]
+    InternalError(String),
+    InvalidTxData(String),
+    SessionError(String),
+    WalletConnectError(WalletConnectError),
+}
+
+impl From<WalletConnectError> for EthWalletConnectError {
+    fn from(value: WalletConnectError) -> Self { Self::WalletConnectError(value) }
+}
+
+/// Eth Params required for constructing WalletConnect transaction.
+pub struct WcEthTxParams<'a> {
+    pub(crate) gas: U256,
+    pub(crate) nonce: U256,
+    pub(crate) data: &'a [u8],
+    pub(crate) my_address: H160,
+    pub(crate) action: Action,
+    pub(crate) value: U256,
+    pub(crate) gas_price: Option<U256>,
+}
+
+impl<'a> WcEthTxParams<'a> {
+    /// Construct WalletConnect transaction json from from `WcEthTxParams`
+    fn prepare_wc_tx_format(&self) -> MmResult<serde_json::Value, EthWalletConnectError> {
+        let mut tx_json = json!({
+            "nonce": u256_to_hex(self.nonce),
+            "from": format!("{:x}", self.my_address),
+            "gas": u256_to_hex(self.gas),
+            "value": u256_to_hex(self.value),
+            "data": format!("0x{}", hex::encode(self.data))
+        });
+
+        if let Some(gas_price) = self.gas_price {
+            tx_json
+                .as_object_mut()
+                .unwrap()
+                .insert("gasPrice".to_string(), json!(u256_to_hex(gas_price)));
```

Comment:
You can use `serde_json::Map` to avoid `as_object_mut().unwrap()` calls for easier mutation and convert it to `serde_json::Value` once you are done with mutating it.